src/server/game/Achievements/AchievementMgr.cpp

```diff
@@ -758,7 +758,22 @@ void AchievementMgr::SendCriteriaUpdate(AchievementCriteriaEntry const* entry, C
         data << uint32(timedCompleted ? 0 : 1); // 1 is for keeping the counter at 0 in client
     data.AppendPackedTime(progress->date);
     data << uint32(timeElapsed);    // time elapsed in seconds
-    data << uint32(0);              // unk
+
+    bool isAverageCriteria = false;
+
+    if ((sAchievementStore.LookupEntry(entry->referredAchievement))->flags & ACHIEVEMENT_FLAG_AVERAGE)
+        isAverageCriteria = true;
+
+    if (AchievementEntryList const* achRefList = sAchievementMgr->GetAchievementByReferencedId(entry->referredAchievement))
+        for (AchievementEntryList::const_iterator itr = achRefList->begin(); itr != achRefList->end(); ++itr)
+            if ((*itr)->flags & ACHIEVEMENT_FLAG_AVERAGE)
+                isAverageCriteria = true;
```

Comment:
Ideally you move this to a function you can call within SendCriteriaUpdate() and BuildAllDataPacket() as it is the same code being repeated.

Reply:
Yes I tried this but was too stupid to get it working. I'll try again.

Reply:
If you need help I can look at it in the weekend